x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportGetRoleMappingsAction.java

```diff
@@ -41,19 +50,43 @@ public TransportGetRoleMappingsAction(
             EsExecutors.DIRECT_EXECUTOR_SERVICE
         );
         this.roleMappingStore = nativeRoleMappingStore;
+        this.clusterStateRoleMapper = clusterStateRoleMapper;
     }
 
     @Override
     protected void doExecute(Task task, final GetRoleMappingsRequest request, final ActionListener<GetRoleMappingsResponse> listener) {
         final Set<String> names;
         if (request.getNames() == null || request.getNames().length == 0) {
-            names = null;
+            names = Set.of();
         } else {
             names = new HashSet<>(Arrays.asList(request.getNames()));
         }
-        this.roleMappingStore.getRoleMappings(names, ActionListener.wrap(mappings -> {
-            ExpressionRoleMapping[] array = mappings.toArray(new ExpressionRoleMapping[mappings.size()]);
-            listener.onResponse(new GetRoleMappingsResponse(array));
+        roleMappingStore.getRoleMappings(names, ActionListener.wrap(nativeRoleMappings -> {
+            final Collection<ExpressionRoleMapping> clusterStateRoleMappings = clusterStateRoleMapper.getMappings(
+                // if the API was queried with a reserved suffix for any of the names, we need to remove it because role mappings are
```

Comment:
Currently, this will fetch from the native mapping store without messing with the name (same as today's behavior) and attempt to fetch from cluster state with/without suffix.

I think that behavior still makes sense: in the unlikely event that users have a native mapping with that suffix it should still be returned via the API. Kibana will indeed get messed up in case there a clashing mappings but I think it's more important that users have a way of getting the clashing native mapping via the API so that they can delete it. 

WDYT about adding a header warning if the native mapping store returns any mappings with a reserved suffix?

Reply:
Sorry, I think I misunderstood the order of things here. 

So, if we have a `mapping-1` the native store and cluster state, and we try to get `mapping-1`, we would get [`mapping-1`, `mapping-1-read-only-operator-mapping`]. 

I think just getting `mapping-1` from the native store would make a little more sense in this case and only return `mapping-1-read-only-operator-mapping` if the full suffix is provided. I think that Kibana probably expects only a single role mapping back when they provide a single name? 

The argument could be made that a user might query for what they have in the operator defined role mappings to validate something and hence use the name from there, so I can see both sides, but if it breaks Kibana, it's probably better to only return from the native store if present in both.

Reply:
So Kibana can handle returning both and will pick the first one. We discussed on slack and decided to stick with the current implementation.

Reply:
Discussed this on Slack:

1. Kibana will pick the first mapping returned for the single mapping view. Since we ensure that native mappings always come first, that's the one that will be chosen.
2. There's (some) value in allowing users to query CS role mappings with or without the suffix. 

As such, keeping this as is.